turbopack/crates/turbopack-core/src/module_graph/merged_modules.rs

```diff
@@ -0,0 +1,584 @@
+        // all modules with the same bitmap.
+        let mut module_merged_groups: FxHashMap<ResolvedVc<Box<dyn Module>>, RoaringBitmapWrapper> =
+            FxHashMap::with_capacity_and_hasher(module_count, Default::default());
+        // Entries that started a new merge group for some deopt reason
+        let mut entry_modules =
+            FxHashSet::with_capacity_and_hasher(module_count, Default::default());
+
+        let mergeable = graphs
+            .iter()
+            .flat_map(|g| g.iter_nodes())
+            .map(async |n| {
+                let module = n.module;
+                let mergeable = ResolvedVc::try_downcast::<Box<dyn MergeableModule>>(module);
+                if let Some(mergeable) = mergeable
+                    && *mergeable.is_mergeable().await?
+                {
+                    return Ok(Some(module));
+                }
+                Ok(None)
+            })
+            .try_flat_join()
+            .await?
+            .into_iter()
+            .collect::<FxHashSet<_>>();
+
+        let mut next_index = 0u32;
+        let visit_count = module_graph
+            .traverse_edges_fixed_point_with_priority(
+                entries.iter().map(|e| (*e, 0)),
+                &mut (),
+                |parent_info: Option<(&'_ SingleModuleGraphModuleNode, &'_ RefData)>,
+                 node: &'_ SingleModuleGraphModuleNode,
+                 _|
+                 -> Result<GraphTraversalAction> {
+                    // On the down traversal, establish which edges are mergable and set the list
+                    // indices.
+                    let (parent_module, hoisted) =
+                        parent_info.map_or((None, false), |(node, ty)| {
+                            (
+                                Some(node.module),
+                                match &ty.chunking_type {
+                                    ChunkingType::Parallel { hoisted, .. } => *hoisted,
+                                    _ => false,
+                                },
+                            )
+                        });
+                    let module = node.module;
+
+                    Ok(if parent_module.is_some_and(|p| p == module) {
+                        // A self-reference
+                        GraphTraversalAction::Skip
+                    } else if let (Some(parent_module), true, true, true) = (
+                        parent_module.filter(|m| {
+                            ResolvedVc::try_downcast::<Box<dyn MergeableModule>>(*m).is_some()
+                        }),
+                        mergeable.contains(&module),
+                        hoisted,
+                        // TODO technically we could merge a sync child into an async parent
+                        !parent_module.is_some_and(|p| async_module_info.contains(&p))
+                            && !async_module_info.contains(&module),
+                    ) {
```

Comment:
```suggestion
                    } else if let Some(parent_module) = 
                        parent_module.filter(|m| {
                            ResolvedVc::try_downcast::<Box<dyn MergeableModule>>(*m).is_some()
                        }) &&
                        mergeable.contains(&module)  &&
                        hoisted  &&
                        // TODO technically we could merge a sync child into an async parent
                        !parent_module.is_some_and(|p| async_module_info.contains(&p))
                            && !async_module_info.contains(&module)
                     {
```

nit: i think `&&` would be a bit easier to read than the `true, true, true`.  it would also trigger short circuiting (though i doubt that is too important